File: internal/detect/monorepo.go
package detect

import (
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

// DetectRepo detects every runnable app in a repo. For a monorepo (turbo / nx /
// pnpm|yarn|npm workspaces / lerna / go.work) it returns one StackFacts per
// member app (with Dir set); otherwise a single root-level stack.
func DetectRepo(root string) []StackFacts {
	rules := loadRules()
	members := monorepoMembers(root)
	var out []StackFacts
	// A repo can be an app at its root AND have workspace members (e.g. a Rails
	// app with a JS streaming service). Include the root only when it's a real
	// app (has a framework) — not when it's just a workspace-root config file.
	if len(members) > 0 {
		if rf, ok := detectWith(rules, root, root); ok && rf.Framework != "" {
			out = append(out, rf)
		}
	}
	for _, m := range members {
		// lockRoot=root: a member's lockfile (bun.lock, uv.lock) usually lives at
		// the repo root, not in the member dir.
		if f, ok := detectWith(rules, filepath.Join(root, m), root); ok {
			f.Dir = m
			out = append(out, f)
		}
	}
	if len(out) > 0 {
		return out
	}
	if f, ok := Detect(root); ok {
		return []StackFacts{f}
	}
	return nil
}

// monorepoMembers returns repo-relative dirs of workspace members that hold a
// detectable app. Empty when the repo is not a monorepo.
func monorepoMembers(root string) []string {
	globs := workspaceGlobs(root)
	if len(globs) == 0 {
		return nil
	}
	// Polyglot monorepos keep apps outside the JS workspace (e.g. a Python
	// backend/ next to a JS frontend/), so also scan conventional app dirs.
	// The "detectable" filter below drops anything that isn't a real app.
	globs = append(globs,
		"backend", "frontend", "api", "server", "web", "client", "worker", "admin",
		"apps/*", "services/*", "packages/*",
	)
	seen := map[string]bool{}
	var dirs []string
	for _, g := range globs {
		matches, _ := filepath.Glob(filepath.Join(root, g))
		for _, m := range matches {
			fi, err := os.Stat(m)
			if err != nil || !fi.IsDir() {
				continue
			}
			rel, err := filepath.Rel(root, m)
			if err != nil || rel == "." || seen[rel] {
				continue
			}
			if f, ok := Detect(m); ok {
				// A frameworkless JS package in a workspace is a library, not a
				// runnable service — skip it (keeps next/vite/nest/... apps only).
				if f.Language == "js" && f.Framework == "" {
					continue
				}
				seen[rel] = true
				dirs = append(dirs, rel)
			}
		}
	}
	sort.Strings(dirs)
	return dirs
}

// workspaceGlobs resolves member globs from whatever monorepo tool the repo uses.
func workspaceGlobs(root string) []string {
	if g := goWorkDirs(root); len(g) > 0 {
		return g
	}
	if g := javaModules(root); len(g) > 0 {
		return g
	}
	if g := pnpmWorkspaceGlobs(root); len(g) > 0 {
		return g
	}
	if g := pkgJSONWorkspaces(root); len(g) > 0 {
		return g
	}
	if g := lernaPackages(root); len(g) > 0 {
		return g
	}
	// turbo/nx declare a monorepo but read members from the underlying workspace;
	// when none was found above, fall back to the conventional layout.
	if exists(root, "turbo.json") || exists(root, "turbo.jsonc") {
		return []string{"apps/*", "packages/*"}
	}
	if exists(root, "nx.json") {
		return []string{"apps/*", "libs/*", "packages/*"}
	}
	return nil
}

func exists(root, name string) bool {
	_, err := os.Stat(filepath.Join(root, name))
	return err == nil
}

func pnpmWorkspaceGlobs(root string) []string {
	data, err := os.ReadFile(filepath.Join(root, "pnpm-workspace.yaml"))
	if err != nil {
		return nil
	}
	var ws struct {
		Packages []string `yaml:"packages"`
	}
	if yaml.Unmarshal(data, &ws) != nil {
		return nil
	}
	return cleanGlobs(ws.Packages)
}

func pkgJSONWorkspaces(root string) []string {
	data, err := os.ReadFile(filepath.Join(root, "package.json"))
	if err != nil {
		return nil
	}
	var pkg struct {
		Workspaces json.RawMessage `json:"workspaces"`
	}
	if json.Unmarshal(data, &pkg) != nil || len(pkg.Workspaces) == 0 {
		return nil
	}
	var arr []string
	if json.Unmarshal(pkg.Workspaces, &arr) == nil {
		return cleanGlobs(arr)
	}
	var obj struct {
		Packages []string `json:"packages"`
	}
	if json.Unmarshal(pkg.Workspaces, &obj) == nil {
		return cleanGlobs(obj.Packages)
	}
	return nil
}

func lernaPackages(root string) []string {
	data, err := os.ReadFile(filepath.Join(root, "lerna.json"))
	if err != nil {
		return nil
	}
	var l struct {
		Packages []string `json:"packages"`
	}
	if json.Unmarshal(data, &l) != nil {
		return nil
	}
	return cleanGlobs(l.Packages)
}

var (
	mavenModuleRe   = regexp.MustCompile(`<module>\s*([^<\s]+)\s*</module>`)
	gradleIncludeRe = regexp.MustCompile(`include\(?\s*['"]:?([^'"]+)['"]`)
)

// javaModules enumerates submodules of a multi-module Maven/Gradle build.
func javaModules(root string) []string {
	var dirs []string
	if data, err := os.ReadFile(filepath.Join(root, "pom.xml")); err == nil {
		for _, m := range mavenModuleRe.FindAllStringSubmatch(string(data), -1) {
			dirs = append(dirs, strings.TrimSpace(m[1]))
		}
	}
	for _, sf := range []string{"settings.gradle", "settings.gradle.kts"} {
		data, err := os.ReadFile(filepath.Join(root, sf))
		if err != nil {
			continue
		}
		for _, m := range gradleIncludeRe.FindAllStringSubmatch(string(data), -1) {
			d := strings.ReplaceAll(strings.TrimSpace(m[1]), ":", "/")
			dirs = append(dirs, strings.TrimPrefix(d, "/"))
		}
	}
	return cleanGlobs(dirs)
}

func goWorkDirs(root string) []string {
	data, err := os.ReadFile(filepath.Join(root, "go.work"))
	if err != nil {
		return nil
	}
	var dirs []string
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		line = strings.TrimPrefix(line, "use")
		line = strings.Trim(line, " (){}")
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "go ") || line == "use" {
			continue
		}
		dirs = append(dirs, strings.TrimPrefix(line, "./"))
	}
	return cleanGlobs(dirs)
}

func cleanGlobs(in []string) []string {
	var out []string
	for _, g := range in {
		g = strings.TrimSpace(g)
		if g == "" || strings.HasPrefix(g, "!") {
			continue
		}
		out = append(out, g)
	}
	return out
}

File: rust/crates/pom_detect/src/monorepo.rs
use std::path::Path;

use pom_config::yaml_node;
use serde_json::Value;

use crate::rules::{detect_with, glob, has_file, load_rules};
use crate::{detect, ResolvedRun, StackFacts};

/// Conventional app folders scanned next to a workspace's own members, for polyglot monorepos.
const CONVENTIONAL: [&str; 11] = [
    "backend",
    "frontend",
    "api",
    "server",
    "web",
    "client",
    "worker",
    "admin",
    "apps/*",
    "services/*",
    "packages/*",
];

fn exists(root: &Path, name: &str) -> bool {
    root.join(name).exists()
}

fn read_json(path: &Path) -> Option<Value> {
    serde_json::from_str(&std::fs::read_to_string(path).ok()?).ok()
}

fn clean(globs: impl IntoIterator<Item = String>) -> Vec<String> {
    globs
        .into_iter()
        .map(|glob| glob.trim().to_string())
        .filter(|glob| !glob.is_empty() && !glob.starts_with('!'))
        .collect()
}

fn string_array(value: Option<&Value>) -> Vec<String> {
    value
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .filter_map(Value::as_str)
                .map(str::to_string)
                .collect()
        })
        .unwrap_or_default()
}

fn go_work_dirs(root: &Path) -> Vec<String> {
    let Ok(text) = std::fs::read_to_string(root.join("go.work")) else {
        return Vec::new();
    };
    clean(text.lines().filter_map(|line| {
        let line = line.trim();
        let line = line.strip_prefix("use").unwrap_or(line);
        let line = line.trim_matches(|c: char| " (){}".contains(c)).trim();
        (!line.is_empty() && !line.starts_with("go ") && line != "use")
            .then(|| line.strip_prefix("./").unwrap_or(line).to_string())
    }))
}

fn java_modules(root: &Path) -> Vec<String> {
    let mut dirs = Vec::new();
    if let Ok(text) = std::fs::read_to_string(root.join("pom.xml")) {
        if let Ok(module) = regex::Regex::new(r"<module>\s*([^<\s]+)\s*</module>") {
            dirs.extend(
                module
                    .captures_iter(&text)
                    .map(|found| found[1].trim().to_string()),
            );
        }
    }
    let include = regex::Regex::new(r#"include\(?\s*['"]:?([^'"]+)['"]"#).ok();
    for file in ["settings.gradle", "settings.gradle.kts"] {
        let (Ok(text), Some(include)) =
            (std::fs::read_to_string(root.join(file)), include.as_ref())
        else {
            continue;
        };
        dirs.extend(include.captures_iter(&text).map(|found| {
            let dir = found[1].trim().replace(':', "/");
            dir.strip_prefix('/').unwrap_or(&dir).to_string()
        }));
    }
    clean(dirs)
}

fn pnpm_globs(root: &Path) -> Vec<String> {
    let Ok(text) = std::fs::read_to_string(root.join("pnpm-workspace.yaml")) else {
        return Vec::new();
    };
    let Ok(Some(document)) = yaml_node::parse(&text) else {
        return Vec::new();
    };
    clean(
        document
            .get("packages")
            .and_then(|packages| packages.items())
            .unwrap_or_default()
            .iter()
            .map(|item| item.text().to_string()),
    )
}

fn package_json_workspaces(root: &Path) -> Vec<String> {
    let Some(package) = read_json(&root.join("package.json")) else {
        return Vec::new();
    };
    let workspaces = package.get("workspaces");
    let listed = string_array(workspaces);
    if !listed.is_empty() {
        return clean(listed);
    }
    clean(string_array(
        workspaces.and_then(|workspaces| workspaces.get("packages")),
    ))
}

fn lerna_packages(root: &Path) -> Vec<String> {
    read_json(&root.join("lerna.json"))
        .map(|lerna| clean(string_array(lerna.get("packages"))))
        .unwrap_or_default()
}

/// Every folder holding an nx `project.json` (depth-bounded, build and dependency trees skipped).
pub(crate) fn nx_projects(root: &Path) -> Vec<String> {
    const SKIP: [&str; 7] = [
        "node_modules",
        ".git",
        "dist",
        "build",
        "coverage",
        "tmp",
        ".nx",
    ];
    let mut found = Vec::new();
    let mut stack = vec![(root.to_path_buf(), 0usize)];
    while let Some((dir, depth)) = stack.pop() {
        let Ok(entries) = std::fs::read_dir(&dir) else {
            continue;
        };
        for entry in entries.flatten() {
            let name = entry.file_name().to_string_lossy().into_owned();
            let path = entry.path();
            if entry.file_type().is_ok_and(|kind| kind.is_dir()) {
                if !SKIP.contains(&name.as_str()) && depth < 4 {
                    stack.push((path, depth + 1));
                }
            } else if name == "project.json" && dir != root {
                if let Ok(relative) = dir.strip_prefix(root) {
                    found.push(relative.to_string_lossy().into_owned());
                }
            }
        }
    }
    found
}

fn workspace_globs(root: &Path) -> Vec<String> {
    for found in [
        go_work_dirs(root),
        java_modules(root),
        pnpm_globs(root),
        package_json_workspaces(root),
        lerna_packages(root),
    ] {
        if !found.is_empty() {
            return found;
        }
    }
    if exists(root, "turbo.json") || exists(root, "turbo.jsonc") {
        return vec!["apps/*".into(), "packages/*".into()];
    }
    if exists(root, "nx.json") {
        let projects = nx_projects(root);
        if !projects.is_empty() {
            return projects;
        }
        return vec!["apps/*".into(), "libs/*".into(), "packages/*".into()];
    }
    Vec::new()
}

/// A package.json with an importable entrypoint (main/module/exports) is a library, not an app.
fn is_js_library(dir: &Path) -> bool {
    read_json(&dir.join("package.json")).is_some_and(|package| {
        ["main", "module"].iter().any(|key| {
            package
                .get(key)
                .and_then(Value::as_str)
                .is_some_and(|value| !value.is_empty())
        }) || package
            .get("exports")
            .is_some_and(|exports| !exports.is_null())
    })
}

fn monorepo_members(root: &Path) -> Vec<String> {
    let mut globs = workspace_globs(root);
    if globs.is_empty() {
        return Vec::new();
    }
    globs.extend(CONVENTIONAL.iter().map(|glob| glob.to_string()));
    let mut members: Vec<String> = Vec::new();
    for pattern in globs {
        for path in glob(root, &pattern) {
            if !path.is_dir() {
                continue;
            }
            let Ok(relative) = path.strip_prefix(root) else {
                continue;
            };
            let relative = relative.to_string_lossy().into_owned();
            if relative.is_empty() || members.contains(&relative) {
                continue;
            }
            if let Some(facts) = detect(&path) {
                if facts.language == "js" && (facts.framework.is_empty() || is_js_library(&path)) {
                    continue;
                }
                members.push(relative);
            }
        }
    }
    members.sort();
    members
}

fn package_dep(dir: &Path, dep: &str) -> bool {
    read_json(&dir.join("package.json")).is_some_and(|package| {
        ["dependencies", "devDependencies"].iter().any(|section| {
            package
                .get(section)
                .and_then(|deps| deps.get(dep))
                .is_some()
        })
    })
}

fn nx_package_manager(root: &Path) -> &'static str {
    [
        ("pnpm-lock.yaml", "pnpm"),
        ("yarn.lock", "yarn"),
        ("bun.lock", "bun"),
        ("bun.lockb", "bun"),
        ("package-lock.json", "npm"),
    ]
    .iter()
    .find(|(file, _)| exists(root, file))
    .map_or("npm", |(_, pm)| pm)
}

fn nx_framework(member: &Path, root: &Path) -> (&'static str, u16) {
    if !glob(member, "next.config.*").is_empty() {
        return ("next", 3000);
    }
    if has_file(member, "nest-cli.json") {
        return ("nest", 3000);
    }
    for dir in [member, root] {
        for (dep, framework, port) in [
            ("@nestjs/core", "nest", 3000),
            ("@angular/core", "angular", 4200),
            ("next", "next", 3000),
            ("express", "express", 3000),
            ("fastify", "fastify", 3000),
        ] {
            if package_dep(dir, dep) {
                return (framework, port);
            }
        }
    }
    if !glob(member, "vite.config.*").is_empty() {
        return ("vite", 5173);
    }
    ("", 0)
}

fn nx_serve(pm: &str, name: &str) -> String {
    match pm {
        "pnpm" => format!("pnpm exec nx serve {name}"),
        "yarn" => format!("yarn nx serve {name}"),
        "bun" => format!("bunx nx serve {name}"),
        _ => format!("npx nx serve {name}"),
    }
}

/// nx apps: deps are hoisted to the root and targets inferred, so each runnable project runs via
/// `nx serve <name>`; libraries and e2e projects are left out.
fn nx_members(root: &Path) -> Vec<StackFacts> {
    let pm = nx_package_manager(root);
    let mut dirs = nx_projects(root);
    if exists(root, "project.json") {
        dirs.push(".".into());
    }
    dirs.sort();
    let mut out = Vec::new();
    for relative in dirs {
        let member = root.join(&relative);
        let Some(project) = read_json(&member.join("project.json")) else {
            continue;
        };
        let name = project
            .get("name")
            .and_then(Value::as_str)
            .unwrap_or_default();
        let project_type = project
            .get("projectType")
            .and_then(Value::as_str)
            .unwrap_or_default();
        let tags = string_array(project.get("tags"));
        let targets = project.get("targets").and_then(Value::as_object);
        let has_target = |target: &str| targets.is_some_and(|targets| targets.contains_key(target));
        let base = Path::new(&relative)
            .file_name()
            .map(|base| base.to_string_lossy().into_owned())
            .unwrap_or_default();
        let e2e = name == "e2e"
            || base == "e2e"
            || name.ends_with("-e2e")
            || base.ends_with("-e2e")
            || has_target("e2e")
            || tags.iter().any(|tag| tag == "type:e2e");
        let runnable = project_type == "application"
            || ["serve", "dev", "start"]
                .iter()
                .any(|target| has_target(target));
        if project_type == "library" || e2e || !runnable {
            continue;
        }
        let name = if name.is_empty() {
            base.clone()
        } else {
            name.to_string()
        };
        let (framework, port) = nx_framework(&member, root);
        out.push(StackFacts {
            dir: if relative == "." {
                String::new()
            } else {
                relative.clone()
            },
            rule_id: "nx".into(),
            language: "js".into(),
            framework: framework.into(),
            package_manager: pm.into(),
            install: format!("{pm} install"),
            run: vec![ResolvedRun {
                kind: "server".into(),
                cmd: nx_serve(pm, &name),
            }],
            setup: Vec::new(),
            port,
            dep_cache: "per-project".into(),
        });
    }
    out
}

/// Every runnable app in a repo: one per monorepo member (with `dir` set), plus the root when it is a real
/// app itself; a single root stack otherwise.
pub fn detect_repo(root: &Path) -> Vec<StackFacts> {
    if exists(root, "nx.json") {
        let members = nx_members(root);
        if !members.is_empty() {
            return members;
        }
    }
    let rules = load_rules();
    let members = monorepo_members(root);
    let mut out = Vec::new();
    if !members.is_empty() {
        if let Some(facts) =
            detect_with(&rules, root, root).filter(|facts| !facts.framework.is_empty())
        {
            out.push(facts);
        }
    }
    for member in members {
        if let Some(mut facts) = detect_with(&rules, &root.join(&member), root) {
            facts.dir = member;
            out.push(facts);
        }
    }
    if !out.is_empty() {
        return out;
    }
    detect(root).into_iter().collect()
}
